Lib/test/test_functools.py

```diff
@@ -2147,6 +2147,43 @@ def __eq__(self, other):
                 return self.arg == other
         self.assertEqual(i("str"), "str")
 
+    def test_method_register(self):
+        class A:
+            @functools.singledispatch(arg=1)
+            def t(self, arg):
+                return "base"
+            @t.register(int)
+            def _(self, arg):
+                return "int"
+            @t.register(str)
+            def _(self, arg):
+                return "str"
```

Comment:
As @ilevkivskyi noted, additional tests are needed to cover the interaction with other method decorators, especially `classmethod`, `staticmethod`, `abstractmethod`, and `partialmethod`.

I think you will find that adding a parameter to the existing `singledispatch` function *isn't* going to be sufficient to solve the problem in the general case (including supporting post-registration of new overloads after the class has been defined). Instead, you'll need something more along the lines of `partialmethod` which is aware of the descriptor protocol, and uses that to delegate to the underlying descriptor's `__get__` method as necessary. While it's a bit long, `singledispatchmethod` would follow the existing convention for descriptor protocol aware variants of previously function-only APIs.

Reply:
Hey @ncoghlan and @ethanhs, I'd been lurking the cpython PRs and had some time today, so I made a quick gist of a possible descriptor-aware implementation of `singlemethoddispatch`. It's far from complete, but it does work for instance, abstract and class methods. (I haven't tried static methods or partial methods yet). I thought I'd link to it here in case it can provide you with some direction / inspiration.

Nice to haves (for my implementation, at least), would be a mechanism which wraps up any methods `register`ed by the decorated method in the same decorators in which that method is wrapped; e.g., if `Foo.bar` is a `classmethod`, then `Foo.bar.register(int, baz)` should make `baz` a class method too. On the other hand, that's complicated to implement and possibly not even the desired behavior, so I'm putting it on the back-burner for now.

Anyway, here's the gist. Heavily inspired by the `partialmethod` implementation, as you'll see: https://gist.github.com/ianhoffman/e25899e8fca3c1ef8772f331b14031d0